### PR TITLE
Added GKE configs for test databases

### PR DESCRIPTION
**Pull request by tarzzz**
- [x] Tested

Part of https://github.com/plotly/streambed/issues/10022

### CODE REVIEW

**Comment by tarzzz** [comment]
@scjody Please review..

The test failures are intermittent and unrelated to these Databases..

File: sample-storage/connections.yaml [hunk 1]
```diff
@@ -85,20 +85,20 @@
     password: w8wfy99DvEmgkBsE
     database: plotly
     port: 50000
-    host: 35.184.35.183
+    host: 35.193.19.132
```
Comment (scjody) [inline sample-storage/connections.yaml]: Can you add hostnames to `test.plotly.host` and use those instead of IP addresses?

Comment (scjody) [reply sample-storage/connections.yaml]: (I suggest setting a TTL of 60 seconds when you do so.)

File: test/docker/ibmdb2/README.md [hunk 2]
```diff
@@ -51,3 +51,25 @@ docker run -ti -p 50000:50000 pdc-db2
...
 To stop the container, just press `CTRL-C`.
+
+
+## GKE Deployment
+
+This directory also contains YAML configurations for deploying this image to GKE cluster.
+To deploy to GKE, run the following command:
+
+```sh
+kubectl create -f db2-deployment.yaml
```
Comment (scjody) [inline test/docker/ibmdb2/README.md]: Please create files that can be applied using `kubectl apply`.

File: test/docker/ibmdb2/README.md [hunk 3]
```diff
@@ -51,3 +51,25 @@ docker run -ti -p 50000:50000 pdc-db2
...
+Confirm that the `db2` pods are running:
+
+```sh
+kubectl get pods
+```
+
+Then create a service using the following command:
+
+```sh
+kubectl create -f db2-service.yaml
```
Comment (scjody) [inline test/docker/ibmdb2/README.md]: Why not create the service right away?

**Comment by scjody** [comment]
Note that this relates to https://github.com/plotly/streambed/issues/10022

@tarzzz As a reminder, please add any related issues to the top comment of any PR you make.

**Comment by tarzzz** [comment]
@scjody: 

  - Moved all the configs to a `deployment/` directory.
  - Added instructions for creating a GKE cluster and running test databases in a README.
  - Created hostnames for test databases on `.test.plotly.host` and updated the code to use the same.
  - Checked that the configurations can be run properly using `kubectl apply`.

Tests are failing because of spark connections (some issue with connectivity to the test db). I am looking into the same..

File: sample-storage/connections.yaml [hunk 4]
```diff
@@ -85,20 +85,20 @@
     password: w8wfy99DvEmgkBsE
     database: plotly
     port: 50000
-    host: 35.184.35.183
+    host: db2.test.plotly.host
```
Comment (scjody) [inline sample-storage/connections.yaml]: 👍

File: test/docker/README.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+
+
+Commands
+========
+
+To Create GKE Cluster:
+
+```sh
+gcloud container clusters create falcon-test-dbs  --zone us-central1-b --machine-type n1-standard-2 --num-nodes 3
```
Comment (scjody) [inline test/docker/README.md]: Please add `--enable-autoupgrade` (this may be a `beta` command)

File: test/docker/README.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+```
+
+To create deployments for test databases:
+
+```sh
+kubectl apply -f deployment/
```
Comment (scjody) [inline test/docker/README.md]: 👍

File: backend/settings.js [hunk 7]
```diff
@@ -59,7 +59,10 @@ const DEFAULT_SETTINGS = {
      * Generally set through the ENV variable:
      * PLOTLY_CONNECTOR_IS_RUNNING_INSIDE_ON_PREM
      */
-    IS_RUNNING_INSIDE_ON_PREM: false
+    IS_RUNNING_INSIDE_ON_PREM: false,
+
+    // Whether it is running in Circle CI environment
+    IS_RUNNING_ON_CIRCLE_CI: false
```
Comment (n-riesco) [inline backend/settings.js]: No need for this. Please, use https://circleci.com/docs/2.0/env-vars/#circleci-environment-variable-descriptions

File: test/docker/spark/Dockerfile [hunk 8]
```diff
@@ -2,7 +2,7 @@ FROM sequenceiq/hadoop-ubuntu:2.6.0
 
 # Get latest java:
 COPY bin/setup_java.sh /setup_java.sh
-RUN chmod +x /setup_java.sh ; /setup_java.sh
+RUN chmod +x /setup_java.sh; sync; /setup_java.sh;
```
Comment (n-riesco) [inline test/docker/spark/Dockerfile]: please, document why `sync` is needed.

Comment (n-riesco) [reply test/docker/spark/Dockerfile]: Oh, I see this is an attempt to fix the failure in circleCI.

File: backend/settings.js (hunk 7 above)
Comment (tarzzz) [reply backend/settings.js]: Circle env vars cannot be used with `getSetting` because the environment variables used in falcon are of the `PLOTLY_CONNECTOR_*` format.

**Comment by tarzzz** [comment]
@scjody : Made some updates on the PR..

  - Uploaded spark and IBM DB2 images publicly to quay.io [here](quay.io/plotly/falcon-test-db2) and [here](quay.io/plotly/falcon-test-spark)

  - IBM DB2 and Spark test databases run inside the circle containers. No more intermittent failures 🎉 
  - Updated spark dockerfile to tackle https://github.com/moby/moby/issues/9547
  - Developers who run test locally can still connect to GKE cluster, so that they dont have to run docker containers for spark and ibm db2

Please let me know if it looks good .. !! 😸

File: backend/settings.js (hunk 7 above)
Comment (n-riesco) [reply backend/settings.js]: There is no need to use `getSetting`, use `process.env.CIRCLECI` instead.

File: test/docker/spark/Dockerfile [hunk 9]
```diff
@@ -1,8 +1,11 @@
 FROM sequenceiq/hadoop-ubuntu:2.6.0
 
+# We are using 'sync' to separate chmod and script execution. See:
+# https://github.com/moby/moby/issues/9547 for details
+
 # Get latest java:
 COPY bin/setup_java.sh /setup_java.sh
-RUN chmod +x /setup_java.sh ; /setup_java.sh
+RUN chmod +x /setup_java.sh; sync; /setup_java.sh;
```
Comment (n-riesco) [inline test/docker/spark/Dockerfile]: My understanding of https://github.com/moby/moby/issues/9547#issuecomment-337684598 is that the `sync` workaround may fail randomly.

I think we should get rid of `chmod` and simply run `sh /setup_java.sh` or `bash /setup_java.sh`.

File: test/docker/spark/Dockerfile [hunk 10]
```diff
@@ -20,7 +23,7 @@ COPY conf/my.cnf /etc/
 
 # Install mysql as hive metastore
 COPY setup_mysql.sh /setup_mysql.sh
-RUN chmod +x /setup_mysql.sh && /setup_mysql.sh;
+RUN chmod +x /setup_mysql.sh; sync; /setup_mysql.sh;
```
Comment (n-riesco) [inline test/docker/spark/Dockerfile]: ditto

File: test/docker/spark/Dockerfile [hunk 11]
```diff
@@ -39,12 +42,12 @@ COPY conf/* $SPARK_HOME/conf/
 COPY bin/* /bin/
 
 # Setup Livy
-RUN chmod +x /bin/setup_livy.sh; /bin/setup_livy.sh;
+RUN chmod +x /bin/setup_livy.sh; sync; /bin/setup_livy.sh;
```
Comment (n-riesco) [inline test/docker/spark/Dockerfile]: ditto

File: test/docker/spark/Dockerfile [hunk 12]
```diff
@@ -39,12 +42,12 @@ COPY conf/* $SPARK_HOME/conf/
...
 COPY conf/livy.conf /livy/conf/
 
 # Workaround for mysql
 RUN echo "skip-grant-tables" | tee -a /etc/my.cnf && service mysql restart; exit 0;
 
 EXPOSE 8998
 
-ENTRYPOINT /etc/bootstrap.sh && service mysql start && chmod +x /bin/setup_database.sh && /bin/setup_database.sh && /livy/bin/livy-server;
+ENTRYPOINT /etc/bootstrap.sh && service mysql start && chmod +x /bin/setup_database.sh && sync && /bin/setup_database.sh && /livy/bin/livy-server;
```
Comment (n-riesco) [inline test/docker/spark/Dockerfile]: ditto

**Comment by n-riesco** [comment]
TODO
- [ ] update the scripts `docker:db2:build` and `docker:db2:start`
- [ ] add similar scripts for spark

I've noticed there is Dockerfile for impala but this PR doesn't include use it in CircleCI.
Is this on purpose?

**Comment by n-riesco** [comment]
I've tried to use the spark DB at spark.test.plotly.host but it fails. See screenshot:

image

**Comment by n-riesco** [comment]
The DB2 sample server at db2.test.plotly.host works for me:

image

File: test/docker/spark/Dockerfile (hunk 9 above)
Comment (tarzzz) [reply test/docker/spark/Dockerfile]: We can't get rid of a command which is necessary.

File: backend/settings.js (hunk 7 above)
Comment (tarzzz) [reply backend/settings.js]: I think we should be consistent in our usage. We **dont** use `process.env` anywhere in code for accessing environment variables.

Going forward if you intend to use `process.env` for accessing environment variables, please open a separate issue where we can get all environment variables use `process.env.*`.

**Comment by tarzzz** [comment]
> update the scripts docker:db2:build and docker:db2:start

The script does not need to be updated as we are not updating the `Dockerfile` for DB2. Please let me know in case I am missing something.

>  add similar scripts for spark

We don't need it because users or developers are not required to run these docker containers on their machine.

**Comment by tarzzz** [comment]
> I've noticed there is Dockerfile for impala but this PR doesn't include use it in CircleCI.
Is this on purpose?

I could not get it working on circle due to some issues. Since it is not a blocker for this PR (I only added  spark and db2 to get tests passing), I will create a separate issue to handle this. Moreover Impala tests do not fail intermittently so we should be good meanwhile.

File: test/docker/spark/Dockerfile (hunk 9 above)
Comment (n-riesco) [reply test/docker/spark/Dockerfile]: Why is it necessary?

Comment (tarzzz) [reply test/docker/spark/Dockerfile]: because `sh setup_java.sh` will also not work without changing the mode (as I recall, I have already tried those options while writing the Dockerfile). 

> My understanding of moby/moby#9547 (comment) is that the sync workaround may fail randomly.

I am pretty positive it will work (for our use case). We can always get back on it if it causes us issues in future.

**Comment by n-riesco** [comment]
@tarzzz 

> The script does not need to be updated as we are not updating the Dockerfile for DB2. Please let me know in case I am missing something.

NVM, I'll do it myself.

> We don't need it because users or developers are not required to run these docker containers on their machine.

Once those sample DB go away, developers will have to run those docker containers. Note that the spark sample DB is failing at  at spark.test.plotly.host the moment.

> I could not get it working on circle due to some issues. Since it is not a blocker for this PR (I only added spark and db2 to get tests passing), I will create a separate issue to handle this. Moreover Impala tests do not fail intermittently so we should be good meanwhile.

Note that we've had to disable tests in datastores.routes.spec.js for impala.

**Comment by n-riesco** [comment]
The impala sample server at impala.test.plotly.host works for me:

image

File: backend/settings.js (hunk 7 above)
Comment (n-riesco) [reply backend/settings.js]: > I think we should be consistent in our usage. We dont use process.env anywhere in code for accessing environment variables.

That's not true.
`process.env` is used by both the backend and the frontend.
Note that it's also used by `webpack.config.*`.

`getSetting` is only used to access properties in `~/.plotly/connector/settings.yaml` and ENV variables starting with `PLOTLY_CONNECTOR`.

For simplicity's sake, please, implement this change as requested.

Comment (tarzzz) [reply backend/settings.js]: No, We don't. We only use it in `webpack.config.*` because we cannot access `getSettings` there. Here is a grep excluding `webpack.*` and compiled bundles:

```
 ✝  ~/plotly/falcon-sql-client   gke-deployments±  ack 'process.env' -r .
app/store/configureStore.js
1:if (process.env.NODE_ENV === 'production') {

test/backend/settings.spec.js
16:        delete process.env.PLOTLY_CONNECTOR_STORAGE_PATH;
21:        delete process.env.PLOTLY_CONNECTOR_STORAGE_PATH;
32:        process.env.PLOTLY_CONNECTOR_STORAGE_PATH = __dirname;
74:        process.env.PLOTLY_CONNECTOR_ACCESS_TOKEN_EXPIRY = String(now);

.eslintrc
55:        "no-process-env": [0],

webpack.config.headless.js
27:            'process.env.NODE_ENV': JSON.stringify('production')

backend/menus.js
86:            submenu: (process.env.NODE_ENV === 'development') ? [{
153:            submenu: (process.env.NODE_ENV === 'development') ? [{

backend/settings.js
158: * Load settings from process.env prefixed with `PLOTLY_CONNECTOR_`
165:    } else if (has(`PLOTLY_CONNECTOR_${settingName}`, process.env)) {
166:        const envObject = process.env[`PLOTLY_CONNECTOR_${settingName}`];
```

Note that we only check for `NODE_ENV`. The only other place is `settings.spec.js` which is used to modify how `getSetting` behaves in tests.

File: test/docker/spark/Dockerfile (hunk 9 above)
Comment (n-riesco) [reply test/docker/spark/Dockerfile]: Why do you think it will work (despite the explanation in https://github.com/moby/moby/issues/9547#issuecomment-337684598)?

File: backend/settings.js (hunk 7 above)
Comment (n-riesco) [reply backend/settings.js]: > No, We don't.

The results you're quoting show that `process.env` is also used in `app/store/configureStore.js`, `backend/menus.js`. 


It's possible to use github to see that it's also used in `backend/main.development.js` and `backend/headless.development.js`. See https://github.com/plotly/falcon-sql-client/search?utf8=%E2%9C%93&q=process.env&type= :

image

image

image

image

image

Comment (n-riesco) [reply backend/settings.js]: Anyway, as I explained above `getSetting` isn't meant to access the ENV variable `CIRCLECI`.

The ENV variable `CIRCLECI` is set automatically by CircleCI and we can access it simply by reading `process.env.CIRCLECI`.

What you're proposing here is that:
1. we add the seeting `IS_RUNNING_ON_CIRCLE_CI` to `backend/settings.js`
2. we set `PLOTLY_CONNECTOR_IS_RUNNING_ON_CIRCLE_CI` in `circle.yml`
3. and we read the value by calling `getSetting('IS_RUNNING_ON_CIRCLE_CI')`

Please, for the sake of simplicity, implement the requested change; i.e. just use `process.env.CIRCLECI` to read the value.

File: test/docker/spark/Dockerfile (hunk 9 above)
Comment (scjody) [reply test/docker/spark/Dockerfile]: I suggest trying the workaround in https://github.com/moby/moby/issues/9547#issuecomment-337684598 . It's more likely to work, and the downside (that we create an extra layer) isn't a big deal for test systems.

If that doesn't work for some reason let's stick with `sync` since it's better than nothing. It's unfortunate how terrible Docker is but that's something we need to deal with.

**Comment by scjody** [comment]
@tarzzz Thanks for the changes. This looks good to me now but you're going to have to work out the `PLOTLY_CONNECTOR_IS_RUNNING_ON_CIRCLE_CI` issue with @n-riesco, and it would be good to try the other workaround for the Docker issue.

File: backend/settings.js (hunk 7 above)
Comment (scjody) [reply backend/settings.js]: Using `getSetting` looks like a clearer and more consistent solution than sprinkling `process.env` reads throughout the code. It's best to centralize configuration handling in one place with one mechanism, and `getSetting` does that.

I support reading `process.env.CIRCLECI` in order to determine if the setting should be set, but it sounds like there are good arguments for doing it the way @tarzzz has implemented it now, namely that it's consistent with the way the other settings are handled.

Comment (n-riesco) [reply backend/settings.js]: @scjody I strongly disagree. `getSetting` isn't meant to access properties like 
 `process.env.CIRCLECI` (as I explained above).

Comment (scjody) [reply backend/settings.js]: @n-riesco I don't see any technical reason why `getSetting` couldn't be enhanced to deal with `process.env.CIRCLECI` as well as environment variables beginning with `PLOTLY_CONNECTOR_`. I suggested it because if you don't want to add another environment variable then that provides a way forward while still using `getSetting`.

`getSetting` is the way we handle settings in the backend of this app. If you don't want to use it with `process.env.CIRCLECI` then let's just add another environment variable as @tarzzz is proposing.

Sprinkling `process.env.CIRCLECI` checks throughout the code will result in something that's significantly less maintainable since you're adding a second configuration mechanism. (There is one `process.env` user in the backend as you pointed out, but it predates the existence of `getSetting` and should probably be changed for consistency. But that's a separate concern.)

Comment (n-riesco) [reply backend/settings.js]: @scjody 

> I don't see any technical reason why getSetting couldn't be enhanced to deal with process.env.CIRCLECI as well as environment variables beginning with PLOTLY_CONNECTOR_.

Technically it's possible; from a design point of view, it is undesirable. I've posted a number of reasons and none of them hasn't been refuted yet (I'll list them again at this end of this comment).

> getSetting is the way we handle settings in the backend of this app. If you don't want to use it with `process.env.CIRCLECI` then let's just add another environment variable as @tarzzz is proposing.

The backend doesn't need to know about `process.env.CIRCLECI`, only the specs do.

> Sprinkling process.env.CIRCLECI checks throughout the code will result in something that's significantly less maintainable since you're adding a second configuration mechanism. 

As I showed in my previous comments, the code is already sprinkled with `process.env`. I agree with you that `process.env` should be removed from the code (but this will have to wait until I have a chance to work again on the `webpack.config.*.js`).

Anyway, this argument doesn't apply to `process.env.CIRCLECI`, because it'd be used only inside the tests.

---

To sum up, here's the list of arguments against introducing the setting `IS_RUNNING_ON_CIRCLE_CI`:
1. The simplest solution (that requires no changes in the code) is to use `process.env.CIRCLECI`. Introducing `IS_RUNNING_ON_CIRCLE_CI` requires changes in `backend/settings.js`.
2. `CIRCLECI` is set automatically by CircleCI  (`PLOTLY_CONNECTOR_IS_RUNNING_ON_CIRCLE_CI` would require to change `circle.yml`)
3. Falcon doesn't need to know about `process.env.CIRCLECI` (only the tests do).

**Comment by tarzzz** [comment]
Updates made:

  - Updated Spark Dockerfile as per the comments.
  - Updated code to use `process.env` instead of `getSetting`
  - Updated GKE cluster to use 3 nodes instead of 2. 

After updating the nodes to 3, spark connector is working for me:

image

@n-riesco : Please let me know if it looks okay.. !!

**Comment by n-riesco** [comment]
@tarzzz I can't test it today, but LGTM. :dancer:

**Comment by n-riesco** [comment]
Just checked the spark server and it works for me too (although it takes a very long time to connect).
